tsdb/engine/tsm1/compact.go

```diff
@@ -578,11 +573,15 @@ func (c *DefaultPlanner) Plan(lastWrite time.Time) ([]CompactionGroup, int64) {
 	// each generation in descending break once we see a file less than 4.
 	end := 0
 	start := 0
-	for i, g := range generations {
-		if g.level() <= 3 {
+	// Walk the generations in reverse to find the last level 4 generation
+	// The backwards iteration is to keep nested lower level generations from preventing
+	// compactions of higher level, later generations.
+	// Test: https://goplay.tools/snippet/t5qSsHiaWqW
+	for i := len(generations) - 1; i >= 0; i-- {
+		if generations[i].level() >= 4 {
+			end = i + 1
 			break
 		}
-		end = i + 1
 	}
```

Comment:
Do we need to check `isInUse` when scanning through the generations? We haven't done anything to filter the in-use generations out yet.